doc/AEPs/aep-0002.md

```diff
@@ -0,0 +1,200 @@
+    AEP: 2
+    Title: A complete develop/publish workflow for Alire crates
+    Author: Alejandro R. Mosteo <amosteo@unizar.es>
+    Status: Draft
+    Created: 06-jul-2020
+
+Abstract
+========
+
+Until date, we do not offer guidelines for a complete closed loop on how to use
+Alire to develop/publish/update a crate. Likewise, we do not have official
+guidelines on how to distribute, outside of the community index, a project that
+uses Alire to obtain its dependencies. 
+
+This document discusses a possible solution for both the closed-loop lifecycle
+of a crate, and the open-loop distribution of projects/releases that are
+themselves not (yet) indexed in Alire.
+
+Rationale
+=========
+
+The root of the problem
+-----------------------
+
+The file `${crate}/alire/crate.toml` (henceforth, the manifest file) is
+currently regenerated from index information during `alr get`, and rewritten
+when dependencies are modified via `alr with`. This is a lossy
+process (although this could be fixed) in which the manifest becomes
+platform-dependent. Hence, we advice against including this
+file in version control, and we ask for it to be excluded from sources
+submitted to the community index.
+
+Even if this file regeneration were not lossy, machine-generated output mangles the
+original formatting. Even with a good pretty-printer for TOML, any comments in
+the original file would be lost. This goes against maintainers, that have to
+track this manual file somehow, yet exclude it from submissions.
+
+On the other hand, developers publishing code outside of Alire, with
+dependencies from Alire, could distribute the manifest file in place and ask
+the users to run `alr update` to obtain all dependencies. This is currently
+frowned upon, again because this can only work today for manifests that are
+platform-independent.
+
+Proposal
+--------
+
+The key change is to fully embrace manual edition of the manifest, and put it
+under version control, in a similar vein to other package managers like
+`cargo`.
+
+The only caveat concerns (at the moment) `alr with`, which is leveraged to
+comfortably make dependencies available in one step (and behind the scenes
+updating the manifest and the project file(s). Thus, there is a need to keep
+`alr with` functionality without regenerating the manifest file, to respect
+user's manual editions.  To meet both ends, the proposal is to add a new
+`[private]` top-level table, at the end of the manifest, properly identified
```

Comment:
If we change `depends-on` to be a [TOML array of table](https://toml.io/en/v1.0.0-rc.1#array-of-tables): `[[depends-on]]`, `alr with` can just append to the manifest:
```
$ alr with gnatcoll
```
```toml
[[depends-on]]
gnatcoll = "*"
```
```
$ alr with apdf
```
```toml
[[depends-on]]
gnatcoll = "*"
[[depends-on]]
apdf = "*"
```
This makes `alr with` very simple to implement and doesn't require a special "do not touch" section in the manifest.
It also means that we cannot have `alr with --del apdf`, but I don't this it is a huge problem.

Reply:
Interesting idea. I need to think how that would impact current dynamic expressions.

I don't like that one can end with dependencies scattered over the manifest, and the "redundant" `depends-on`. OTOH I don't like having dependencies split between the manifest and the lockfile or private part either. This looks cleaner.